File: src/detect/features.ts
import { join } from 'node:path';
import { exists } from '../utils/fs.js';

export interface FeatureFlags {
	git: boolean;
	ci: 'github-actions' | 'gitlab-ci' | 'jenkins' | null;
	docker: boolean;
	playwright: boolean;
	semgrep: boolean;
	firebase: boolean;
	vercel: boolean;
}

export async function detectFeatures(cwd: string): Promise<FeatureFlags> {
	const [git, githubActions, gitlabCi, jenkinsfile, dockerfile, dockerCompose, playwrightConfig, semgrepConfig, firebase, vercel] =
		await Promise.all([
			exists(join(cwd, '.git')),
			exists(join(cwd, '.github', 'workflows')),
			exists(join(cwd, '.gitlab-ci.yml')),
			exists(join(cwd, 'Jenkinsfile')),
			exists(join(cwd, 'Dockerfile')),
			exists(join(cwd, 'docker-compose.yml')),
			exists(join(cwd, 'playwright.config.ts')).then(
				(e) => e || exists(join(cwd, 'playwright.config.js')),
			),
			exists(join(cwd, '.semgrep.yml')),
			exists(join(cwd, 'firebase.json')),
			exists(join(cwd, 'vercel.json')),
		]);

	let ci: FeatureFlags['ci'] = null;
	if (githubActions) ci = 'github-actions';
	else if (gitlabCi) ci = 'gitlab-ci';
	else if (jenkinsfile) ci = 'jenkins';

	return {
		git,
		ci,
		docker: dockerfile || dockerCompose,
		playwright: playwrightConfig,
		semgrep: semgrepConfig,
		firebase,
		vercel,
	};
}

File: internal/detect/features.go
package detect

import (
	"github.com/samahlstrom/forge-cli/internal/util"
	"path/filepath"
)

// FeatureFlags indicates which CI/CD and tooling features are present.
type FeatureFlags struct {
	Git        bool   `json:"git"`
	CI         string `json:"ci,omitempty"` // github-actions, gitlab-ci, jenkins, or ""
	Docker     bool   `json:"docker"`
	Playwright bool   `json:"playwright"`
	Semgrep    bool   `json:"semgrep"`
	Firebase   bool   `json:"firebase"`
	Vercel     bool   `json:"vercel"`
}

// DetectFeatures checks for CI, Docker, and other tooling.
func DetectFeatures(cwd string) FeatureFlags {
	f := FeatureFlags{
		Git:        util.Exists(filepath.Join(cwd, ".git")),
		Docker:     util.Exists(filepath.Join(cwd, "Dockerfile")) || util.Exists(filepath.Join(cwd, "docker-compose.yml")),
		Playwright: util.Exists(filepath.Join(cwd, "playwright.config.ts")) || util.Exists(filepath.Join(cwd, "playwright.config.js")),
		Semgrep:    util.Exists(filepath.Join(cwd, ".semgrep.yml")),
		Firebase:   util.Exists(filepath.Join(cwd, "firebase.json")),
		Vercel:     util.Exists(filepath.Join(cwd, "vercel.json")),
	}

	switch {
	case util.Exists(filepath.Join(cwd, ".github", "workflows")):
		f.CI = "github-actions"
	case util.Exists(filepath.Join(cwd, ".gitlab-ci.yml")):
		f.CI = "gitlab-ci"
	case util.Exists(filepath.Join(cwd, "Jenkinsfile")):
		f.CI = "jenkins"
	}

	return f
}
